dsl/src/main/scala/org/http4s/dsl/Path.scala

```diff
@@ -249,6 +249,48 @@ abstract class OptionalQueryParamDecoderMatcher[T: QueryParamDecoder](name: Stri
     ).toOption
 }
 
+/**
+  * Capture a query parameter that appears 0 or more times.
+  *
+  * {{{
+  *   case class Foo(i: Int)
+  *   implicit val fooDecoder: QueryParamDecoder[Foo] = ...
+  *   implicit val fooParam: QueryParam[Foo] = ...
+  *
+  *   object FooMatcher extends OptionalMultiQueryParamDecoderMatcher[Foo]("foo")
+  *   val service: HttpService = {
+  *     // matches http://.../closest?foo=2&foo=3&foo=4
+  *     case GET -> Root / "closest" :? FooMatcher(Some(Seq(2,3,4))) => ...
+  *
+  *     /*
+  *     *  matches http://.../closest?foo=2&foo=3&foo=4 as well as http://.../closest (no parameters)
+  *     *  or http://.../closest?foo=2 (single occurrence)
+  *     */
+  *     case GET -> Root / "closest" :? FooMatcher(is) => ...
+  * }}}
+  */
+abstract class OptionalMultiQueryParamDecoderMatcher[T: QueryParamDecoder](name: String) {
+  import scalaz.{ValidationNel, Success, Failure}
+
+  def unapply(params: Map[String, Seq[String]]): Option[Option[ValidationNel[ParseFailure, Seq[T]]]] = {
+    params.get(name) match {
+      case None => Some(None) // absent
+      case Some(values) => {
+        val parses: Seq[ValidationNel[ParseFailure, T]] = values.map(s => QueryParamDecoder[T].decode(QueryParameterValue(s)))
+        val parsed: ValidationNel[ParseFailure, Seq[T]] = parses.foldLeft(Success(Seq[T]()) : ValidationNel[ParseFailure, Seq[T]])((
```

Comment:
I feel like there's a `traverse` in here, but as long as the return type is right, I'm not that worried about figuring it out right now.  I'd rather get 2.12 in people's hands.